[BottomNavigation] Lays out the bar when titleVisibility is changed

Ruberik opened the pull request:
Closes #6151.

Ruberik committed:
[BottomNavigation] Lays out the bar immediately when bottomNav.titleVisibility is changed.

Ruberik commented:
FYI, I just updated this change so that the demo is in its own file, instead of in BottomNavigationTypicalUseExample.

romoore commented on components/BottomNavigation/src/private/MDCBottomNavigationItemView.m:
```diff
@@ -383,6 +383,7 @@ - (void)setTitle:(NSString *)title {
 -(void)setTitleVisibility:(MDCBottomNavigationBarTitleVisibility)titleVisibility {
   _titleVisibility = titleVisibility;
   [self updateLabelVisibility];
+  [self centerLayoutAnimated:YES];
```
Does it not work to call `setNeedsLayout` here instead so any layout related to visibility can be performed in a standard UIKit layout pass?

Ruberik replied:
Yes, that's better. I'm actually going to move that into updateLabelVisibility, since that seems like a better place for that to live.

romoore replied:
Sounds good! 👍

romoore reviewed (changes requested):
Thanks for the PR! Is there a unit test that could be added to test this condition? If not, is there a small snippet that could be turned into a snapshot test?  Perhaps changing the bounds/frame while the titles are not visible?

It would be great to find the code path that causes this bug and make sure we've gotten it squashed.

material-automation commented:
bazel detected changes to the following targets:

```
//components/BottomNavigation:unit_tests
//components/BottomNavigation:unit_tests_IPAD_PRO_12_9_IN_9_3
//components/BottomNavigation:unit_tests_IPAD_PRO_12_9_IN_9_3_test_bundle
//components/BottomNavigation:unit_tests_IPHONE_5_IN_8_1
//components/BottomNavigation:unit_tests_IPHONE_5_IN_8_1_test_bundle
//components/BottomNavigation:unit_tests_IPHONE_7_PLUS_IN_10_3
//components/BottomNavigation:unit_tests_IPHONE_7_PLUS_IN_10_3_test_bundle
//components/BottomNavigation:unit_tests_IPHONE_X_IN_11_0
//components/BottomNavigation:unit_tests_IPHONE_X_IN_11_0_test_bundle
```

<!-- identifier: affected-targets -->

material-automation commented:
The following files are not in any BUILD targets:

```
components/BottomNavigation/examples/BottomNavigationTitleVisibilityChangeExample.swift
```


<!-- identifier: bazel-rule-coverage -->

Ruberik committed:
Adds a test, and switches to a better-placed setNeedsLayout.

Ruberik committed:
Changes the name of the new example to better match convention. It's too long, but only "Swift" gets cut off, and it's searchable.

Ruberik commented:
@romoore Test added! I'll consider this a helpful introduction to Objective-C; all the hair will grow back.

Ruberik commented:
BTW, while I have your attention on this test file, does testSetSelectedItemTintColorUpdatesInkColor look right to you? It has the same check for item1 and item2, even though one is selected, the other isn't, and the point of the test is to make sure selection changes something.

romoore commented:
Hi @Ruberik . That test changes the `selectedItemTintColor` on both a selected and unselected item. It then confirms that both have changed.  Since how the Ripple (here called "ink") color's actual value is documented publicly, the test doesn't make any assumptions about what the new color should be - only that it isn't the original color.  The reason for that test (and the related fix in https://github.com/material-components/material-components-ios/pull/4950 was to fix a bug in the implementation and update how Ripple color is applied to Bottom Navigation items.

I could certainly see an improvement being made where we assert that both items have the same (changed) color. I've opened https://github.com/material-components/material-components-ios/pull/6190 to add that check.

Was there something else in the test that you think could be improved?

romoore commented on components/BottomNavigation/examples/BottomNavigationTitleVisibilityChangeExample.swift:
```diff
@@ -0,0 +1,117 @@
+  override func viewWillLayoutSubviews() {
+    super.viewWillLayoutSubviews()
+    layoutBottomNavBar()
+    addInstructionLabel()
```
I may be mistaken, but this looks like it creates a new UILabel every time the view is resized or rotated?   Did you mean to put the call to `addInstructionLabel()` in `viewDidLoad`?

Ruberik replied:
Thank you for spotting that. Moved to viewDidLoad().

romoore commented on components/BottomNavigation/examples/BottomNavigationTitleVisibilityChangeExample.swift:
```diff
@@ -0,0 +1,117 @@
+  }
+  
+  func commonBottomNavigationTypicalUseSwiftExampleInit() {
+    view.backgroundColor = .lightGray
```
Nit: Please use `colorScheme.backgroundColor` for the background color.

Ruberik replied:
Fixed, here and in other BottomNav Swift examples.

romoore reviewed (approved):
LGTM. Just a pair of minor comments about the example.

material-automation commented:
The following files are not in any BUILD targets:

```
components/BottomNavigation/examples/BottomNavigationTitleVisibilityChangeExample.swift
```


<!-- identifier: bazel-rule-coverage -->

randallli commented:
I reran the kokoro CI.
We need to merge develop so that we get #6184 which should fix the failing kokoro/pr/website check.

randallli commented:
error from kokoro (Internal CI):

Test Suite 'BottomNavigationItemViewTests' started at 2019-01-07 07:58:32.626
Test Case '-[BottomNavigationItemViewTests testBadgeCenterIsCorrectWithoutRTL]' started.
Test Case '-[BottomNavigationItemViewTests testBadgeCenterIsCorrectWithoutRTL]' passed (0.002 seconds).
Test Case '-[BottomNavigationItemViewTests testBadgeCenterIsCorrectWithRTL]' started.
Test Case '-[BottomNavigationItemViewTests testBadgeCenterIsCorrectWithRTL]' passed (0.002 seconds).
Test Case '-[BottomNavigationItemViewTests testContentInsetLayout]' started.
Test Case '-[BottomNavigationItemViewTests testContentInsetLayout]' passed (0.016 seconds).
Test Case '-[BottomNavigationItemViewTests testHorizontalMarginLayout]' started.
Test Case '-[BottomNavigationItemViewTests testHorizontalMarginLayout]' passed (0.003 seconds).
Test Case '-[BottomNavigationItemViewTests testSetSelectedItemTintColorUpdatesInkColor]' started.
Test Case '-[BottomNavigationItemViewTests testSetSelectedItemTintColorUpdatesInkColor]' passed (0.002 seconds).
Test Case '-[BottomNavigationItemViewTests testSetTitleVisibilityUpdatesLayout]' started.
components/BottomNavigation/tests/unit/BottomNavigationItemViewTests.m:141: error: -[BottomNavigationItemViewTests testSetTitleVisibilityUpdatesLayout] : ((imageMidY) equal to (32.75) +/- (0.001)) failed: ("32.5") is not equal to ("32.75") +/- ("0.001")

romoore commented on components/BottomNavigation/tests/unit/BottomNavigationItemViewTests.m:
```diff
@@ -127,6 +127,28 @@ - (void)testSetSelectedItemTintColorUpdatesInkColor {
+  view.titleVisibility = MDCBottomNavigationBarTitleVisibilityAlways;
+  [view layoutIfNeeded];
+  CGFloat imageMidY = CGRectGetMidY(view.iconImageView.frame);
+  XCTAssertEqualWithAccuracy(imageMidY, 32.75, 0.001);
```
Tip: Consider making this a less-than-equal and greater-than-equal so you can verify the "approximate" location.  It seems like we don't care too much about the exact placement, but rather only that it is "to the left of center" by some reasonable amount.

Ruberik replied:
Good point. I appreciate you giving a broader view; I admit I was happy just to get a passing test that wouldn't have passed before.

Ruberik committed:
Changes background colours to follow the color scheme, and moves a label's creation into viewDidLoad().

Ruberik committed:
Makes testSetTitleVisibilityUpdatesLayout less implementation-specific.

Ruberik commented:
I'll preface this by saying that I'm a Git noob who, so far, has managed to make things... sort of work.

I've made the changes, but I can't seem to get them on github.com/Ruberik/material-components-ios. Here's what happened:

1. When I created this PR, I (unwisely?) made it in my own `develop` branch.
2. When I rebase my local `develop` branch onto material-components' `develop` branch, it works fine. When I try to force push (is that really what I should be doing?) onto github.com/Ruberik/material-components-ios's `develop` branch, here's what I get. The commandline is coming from Sourcetree, for what it's worth:

```
git --no-optional-locks -c color.branch=false -c color.diff=false -c color.status=false -c diff.mnemonicprefix=false -c core.quotepath=false -c credential.helper=sourcetree -c lfs.customtransfer.bitbucket-media-api.path=/Applications/Sourcetree.app/Contents/Resources/git_local/git-lfs/git-lfs-bitbucket-media-api push -v --force origin refs/heads/develop:refs/heads/develop 
Pushing to https://github.com/Ruberik/material-components-ios.git
open /Users/Bartholomew/dev/material-components-ios/snapshot_test_goldens/goldens_64/MDCTextAreaOutlinedControllerBaselineCharacterCountArabicSnapshotTests/testTextFieldEmptyDisabled_11_2@2x.png: no such file or directory
error: failed to push some refs to 'https://github.com/Ruberik/material-components-ios.git'
Completed with errors, see above
```

Sorry to drag you into my own personal Git troubles.

romoore commented:
Can you try  `git lfs fetch --all` and then push?  Let me know if that helps.

romoore commented:
In general, if you found you've created a commit in the wrong branch, you can do the following:

1. Use `git log` to find the commits since you branched from upstream's `develop`. Grab the commit hash.
2. Use `git reset --soft <COMMIT HASH>` to undo ALL of your work and have it sitting as if it's just been modified on the file system.
3. Use `git stash` to put all modified files (but not newly-created files) into the stash
4. Use `git checkout -b my-new-branch` to create the new branch, `my-new-branch`
5. Use `git stash apply` to unstash your changes and put them in the working directory
6. Commit your changes to the new branch
7. Push (forced?) to the existing remote branch or push to a new remote branch (and open a new PR).

As clean-up, you'll also want to do `git stash drop` to drop the latest stash entry. **Be careful!**  If you aren't sure, you can always `git stash list` to take a look at what's in your stash.

romoore commented:
Oops, I had the wrong `git lfs` command above. It's been correct to spell "all" correctly.

romoore commented on components/BottomNavigation/tests/unit/BottomNavigationItemViewTests.m:
```diff
@@ -127,6 +127,27 @@ - (void)testSetSelectedItemTintColorUpdatesInkColor {
+
+  // Then
+  CGFloat imageMidYWithoutTitle = CGRectGetMidY(view.iconImageView.frame);
+  XCTAssertGreaterThan(imageMidYWithoutTitle, imageMidYWithTitle + 1);  // We add 1 to make sure the change in position is greater than a rounding error. Practically, it's likely to be greater than 10.
```
Nit: This line looks a bit long. Consider breaking it up or running clang-format on your changes.

Ruberik replied:
Broken up and clang-formatted.

material-automation commented on components/BottomNavigation/tests/unit/BottomNavigationItemViewTests.m:
```diff
@@ -127,6 +127,27 @@ - (void)testSetSelectedItemTintColorUpdatesInkColor {
+
+  // Then
+  CGFloat imageMidYWithoutTitle = CGRectGetMidY(view.iconImageView.frame);
+  XCTAssertGreaterThan(imageMidYWithoutTitle, imageMidYWithTitle + 1);  // We add 1 to make sure the change in position is greater than a rounding error. Practically, it's likely to be greater than 10.
```
clang-format suggested the following change:


```
-  XCTAssertGreaterThan(imageMidYWithoutTitle, imageMidYWithTitle + 1);  // We add 1 to make sure the change in position is greater than a rounding error. Practically, it's likely to be greater than 10.
+  XCTAssertGreaterThan(
+      imageMidYWithoutTitle,
+      imageMidYWithTitle + 1);  // We add 1 to make sure the change in position is greater than a
+                                // rounding error. Practically, it's likely to be greater than 10.
```

<!-- identifier: clang-format -->
<!-- File: components/BottomNavigation/tests/unit/BottomNavigationItemViewTests.m -->

material-automation commented:
clang-format recommended changes.

Consider installing [git-clang-format](https://github.com/llvm-mirror/clang/blob/master/tools/clang-format/git-clang-format) and running the following command:

```bash
git clang-format $(git merge-base origin/develop HEAD)
```

<!-- identifier: clang-format -->

material-automation commented:
The following files are not in any BUILD targets:

```
components/BottomNavigation/examples/BottomNavigationNilBadges.swift
components/BottomNavigation/examples/BottomNavigationTitleVisibilityChangeExample.swift
components/BottomNavigation/examples/BottomNavigationTypicalUseExample.swift
```


<!-- identifier: bazel-rule-coverage -->

Ruberik committed:
Fixes a comment with git clang-format.

Ruberik commented:
Thanks! That `git lfs` command fixed me up nicely. I believe the currently-pushed thing in my fork is properly rebased and has addressed all comments.

Thanks for all the help writing this. I've definitely become a more competent git user after this.

romoore commented on components/BottomNavigation/examples/BottomNavigationTitleVisibilityChangeExample.swift:
```diff
@@ -0,0 +1,120 @@
+  }
+  
+  override func viewDidLoad() {
+    addInstructionLabel()
```
Did you forget to call `super.viewDidLoad()`?

Ruberik replied:
I did! Did a lint pick that up, or did you spot it?

romoore replied:
That was just me. I've often made the same mistake in the past, so I try to catch it where I can. ;-)

Ruberik committed:
Adds a missing call to super.viewDidLoad().

romoore commented:
;_; Looks like the force-push lost the commit history.  I'll take another review at this today.

Ruberik commented:
Aww! Sorry. Dumb question: is there a better way to handle catching up with your `develop` branch than rebasing my changes in my local repository, then force pushing them? I tried to find a way to directly rebase my GitHub repository, but couldn't find one.

romoore commented:
If you do your work on a local branch, you can then `git merge develop` to pull changes into your development branch.  Here's my typical workflow:

1. `git checkout -b issue-1234`
2. Do work....
3. `git commit`
4. `git push origin issue-1234`
5. `git checkout develop`
6. `git pull upstream develop` (this keeps my local `develop` fast-forwarded with changes upstream)
7. `git push origin develop`
7. `git checkout issue-1234`
8. `git merge develop`  (resolve any conflicts)
9. `git push origin issue-1234`

I will repeat 2-9 for any additional work, feedback, etc.  If a PR is still sitting in GitHub in the morning, I try to do a merge from `develop` before doing any more work.  I just opened https://github.com/material-components/material-components-ios/issues/6280 to reduce the number of errors not caught by outdated PRs.

Ruberik commented:
Thanks for explaining that! All the guides I could find seemed to assume nothing was happening to the repository other than the reader's changes, and skipped 4-10. Is there a Google guide to GitHub workflow somewhere external? Or a good guide somewhere that you used to learn this stuff?

romoore commented:
I'm not sure, but Atlassian has a good description of the differences.

https://www.atlassian.com/git/tutorials/merging-vs-rebasing
